src/core/lib/surface/init.cc

```diff
@@ -193,8 +223,13 @@ void grpc_shutdown_internal_locked(void) {
   grpc_core::ApplicationCallbackExecCtx::GlobalShutdown();
   g_shutting_down = false;
   gpr_cv_broadcast(g_shutting_down_cv);
+  free(g_shutting_down_cv);
   // Absolute last action will be to delete static metadata context.
   grpc_destroy_static_metadata_ctx();
+
+  // Executes all custom cleanup functions, which were registered via
+  // grpc_on_shutdown_callback() and grpc_on_shutdown_callback_with_arg()
+  delete ShutdownData::get();
```

Comment:
In addition to the comments I made in my review, note that the grpc library can be init'ed and shutdown numerous times during the running of an application. It does not need to stay alive for the entire duration of the application, and it can be restarted later. I think doing these sorts of frees on function-static and global variables without actually recreating them one a later init is not going to work for the actual usage allowed by gRPC.

Reply:
I think, then there should be a function
`GRPCAPI void grpc_shutdown_grpc_library();`
to trigger that cleanup. This function can be called at process exit or when unloading a DLL that uses gRPC.